---
title: "RNA-Seq Analysis Training"
author: "Nicol√°s Vera"
link-citations: true
bibliography: references.bib
csl: 3d-printed-materials-and-systems.csl 
date: "05/6/2022"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 3
    theme: united
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T,results = "hide",warning = F,message = F)
```

```{r,warning=FALSE}
library(readxl)
library(dplyr)
library(tidyverse)
library(EnhancedVolcano)
library(pheatmap)
library(dplyr) 
library(msigdbr) 
library(enrichplot)
library(clusterProfiler)
library(GOplot)
library(simplifyEnrichment)
library(ggfortify)
library(plotly)
library(org.Hs.eg.db)
library(ggvenn)
library(gplots)
library(DT)
library(DESeq2)
library(DOSE)
```

```{r}

# Thiscode has been taken from #http://leekgroup.github.io/regionReportSupp/DESeq2-example/index.html
#and turned into a function
interactive_table <- function(results) {

  for(i in which(colnames(results) %in% c('pvalue', 'padj'))) results[, i] <- format(results[, i], scientific = TRUE)


datatable(head(as.data.frame(results), n = 50), options = list(pagingType='full_numbers', pageLength=10, scrollX='100%'), escape = FALSE, rownames = FALSE,caption = htmltools::tags$caption( style = 'caption-side: top; text-align: center; color:black; font-size:200% ;',gsub("standard error: Response ","",results@elementMetadata@listData$description[3]))) %>% formatRound(which(!colnames(results) %in% c('pvalue', 'padj', 'Feature')), 3)

}
```

# Introduction

Tracing its first steps to the second half of the 19th century, immunotherapy has emerged as one of the most promising types of cancer treatment. One of the obstacles that prevents the immune system from attacking cancer cells is overexpression of PD-L1 in these cells. The biological function of the PD-1/PD-L1 axis is negative regulation of immune activation. PD-L1 binds to the PD-1 receptor in the surface of T cells, inducing either T cell exhaustion or apoptosis, thus enabling immune escape by PD-L1 over-expressing cancer cells. PD-1 checkpoint blockade works by blocking the PD-1/PD-L1 axis through the use of monoclonal antibodies such as nivolumab and pembrolizumab [@waldman-2020]. Despite checkpoint blockade therapy's having proved itself effective, chief among the challenges it faces is the high number of non-responsive patients [@nowicki-2018].

For the present project we have a dataset (GEO Accession Number: ) consisting of mRNA-seq data from 28 tumour samples from melanoma patients obtained prior to their treatment with PD-1 blocker pembrolizumab. The samples (n = 28) are divided into complete responders (CR, n = 5), partial responders (PR, n = 10) and non-responders (PD, n = 13). Two of the samples belong to the same patient.

That the samples are pre-treatment is especially useful for finding factors that are predictive of a good response, as well as giving insights about the mechanisms involved in impeding the expected response to the treatment. In this project we will try to find factors that prevent response. Complementarily, we will also try to ascertain whether complete responses and partial responses are just due to the same factors being present in different degrees or whether they represent qualitatively distinct phenomena.

In the first place, we will use DESEq2 to find diferentially expressed genes between the different groups. Then, we will use the results of DESEq2 to find significantly enriched pathways and gene sets between the groups.

Since we have three groups, there will be three pairwise comparisons:

-Complete Response vs Progressive Disease

-Partial Response vs Progressive Disease.

-Complete Response vs Partial Response.

# Results

## PCA

After importing the data, we will plot the two first principal components after applying a variance stabilizing transformation (vst) in order to see whether treatment response explains a considerable amount of the variability between samples and in order to identify potential outliers.

```{r}
#The data and metadata is imported

if(!file.exists("GSE78220-expression.txt")){
  download.file("https://maayanlab.cloud/notebook-generator-server/download?gse=GSE78220&platform=GPL11154&version=v6&content=expression","GSE78220-expression.txt")
}

if(!file.exists("GSE78220-metadata.txt")){
  download.file("https://maayanlab.cloud/notebook-generator-server/download?gse=GSE78220&platform=GPL11154&version=v6&content=metadata","GSE78220-metadata.txt")
}

raw_counts <- read.table("GSE78220-expression.txt",header = T)

metadata2 <- read_xlsx("GSE78220_series_matrix.xlsx",sheet = 2)

metadata <- read.table("GSE78220-metadata.txt",header = T,sep = "\t",row.names = 1)

metadata <- metadata %>%
  #filter(str_detect(resp,"Complete") | str_detect(resp,"Progressive")) %>%
  mutate(Response = as.factor(recode(anti.pd.1.response, 
                    "Complete Response" = "CR",
                    "Progressive Disease" = "PD",
                    "Partial Response" = "PR")),
         )

raw_counts <- raw_counts[,rownames(metadata)]

#We use patient number instead of GEO accession number for readability
colnames(raw_counts) <- metadata$Sample.Title

metadata <- metadata %>%
  remove_rownames() %>%
  column_to_rownames(var = "Sample.Title")
```

```{r,warning=FALSE}
dds <- DESeqDataSetFromMatrix(countData = raw_counts,colData = metadata,design = ~Response)

dds <- DESeq(dds)
```

```{r}
transformed_dds <- vst(dds,blind = T)

pcaData <- plotPCA(transformed_dds, intgroup=c("Response","previous.mapki","patient.id"), returnData=TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=Response, shape=previous.mapki)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + geom_text_repel(aes(label = patient.id))
```

There is no obvious clustering of the different groups. As Pt2 & Pt28 seem to be represent too much of PC2 variance, we will consider them outliers and remove them for downstream analysis.

```{r}
#Pt2 and Pt28 are removed
raw_counts <- raw_counts[,-c(9,27)]

metadata <- metadata[-c(9,27),]
```

After removing them, we calculate and plot again the first two principal components.

```{r,warning=FALSE}
dds <- DESeqDataSetFromMatrix(countData = raw_counts,
                              colData = metadata,design = ~Response)

dds <- DESeq(dds)
```

```{r}
transformed_dds <- vst(dds,blind = T)

pcaData <- plotPCA(transformed_dds, intgroup=c("Response","previous.mapki","patient.id"), returnData=TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=Response, shape=previous.mapki)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + geom_text_repel(aes(label = patient.id))
```

With the two patients removed, the three groups still don't cluster but there are outliers no longer.

## Differentially expressed genes

Let's show the differentially expressed genes in each one of the three comparisons. In any of the comparisons the first group in its name is the numerator and the second group is the denominator (i.e. a gene with a positive log2FC in "CR to PD" is overexpressed in CR in relation to PD).

```{r}

list_results <- list(resCR_PD = c("CR","PD"),resPR_PD = c("PR","PD"),resCR_PR = c("CR","PR"))

results_contrasts <- lapply(list_results,function(i) results(dds,contrast = c("Response",i[1],i[2])))
```

```{r}
lapply(results_contrasts,function(i) interactive_table(i))
```

## Volcano plots

Now we will visualize, through volcano plots, the amount of genes that are significantly differentially expressed (p adjusted \< 0.01) and whose log2foldchange is \> 1 (i.e., a change twofold or more) for each of the three pairwise comparisons.

```{r,warning=FALSE,results='hide'}
lapply(results_contrasts,function(i) EnhancedVolcano(i,lab = rownames(i),x = "log2FoldChange",y="padj",pCutoff = 0.01, title = gsub("standard error: Response ","",i@elementMetadata@listData$description[3])))
```

It catches our attention that the FGA gene is enriched both in PD and in CR with respect to PR. We will next plot a Venn diagram showing the overlap between the genes differentially expressed in each of the three comparisons.

```{r}
siglists <- lapply(results_contrasts,function(i) as.data.frame(i) %>% filter(abs(-log2FoldChange) > 1,padj <= 0.01,baseMean >20) %>%
  rownames_to_column("Gene") %>%
  pull(Gene))

venn(siglists)
```

There doesn't seem to be much overlap between A (CR vs PD) and B (PR vs PD), suggesting that the factors predicting complete and partial response are somewhat different. The degree of overlap between B (PR vs PD) and C (CR vs PR) may not be due to common factors but to genes down-regulated in one of the comparisons and up-regulated in the other,as was the case with FGA in the volcano plot. Let's check that possibility by plotting a Venn diagram exclusively of up-regulated genes and then a down-regulated-exclusive one. Here we will use an extra condition for significance: that the mean expression of the gene is \> 20 (in order to avoid huge ratios between insignificant expression levels).

```{r}
siglists <- lapply(results_contrasts,function(i) as.data.frame(i) %>% filter(log2FoldChange >1 ,padj <= 0.01,baseMean >20) %>%
  rownames_to_column("Gene") %>%
  pull(Gene))

venn(siglists)
title(main = "Upregulated only")
```

```{r}
siglists <- lapply(results_contrasts,function(i) as.data.frame(i) %>% filter(log2FoldChange < -1,padj <= 0.01,baseMean >20) %>%
  rownames_to_column("Gene") %>%
  pull(Gene))

venn(siglists)
title(main = "Downregulated only")
```

As both the diagram that includes only the up-regulated genes and that that includes only the down-regulated show 0 overlap between B and C, all the overlap seen in the first Venn diagram must be due to genes being down-regulated in one of the two comparisons being up-regulated in the other comparison. This is noteworthy, as the two comparisons (PR vs PD and CR vs PR) are in the same direction (a higher degree of response is compared to a lower one), i.e. some genes are up-regulated both in the complete response and in the progressive disease group relative to the partial response group.

## Heatmaps

Now we will explore how much the top 10 differentially over-expressed and under-expressed genes in each comparison are expressed in each of the patients by means of heatmaps.

```{r,results='hide'}
results2heatmap <- function(results,dds){
  
  sig_results_up <- as.data.frame(results) %>%
    rename_with(function(x) gsub(".{8}\\.","",x)) %>%
    filter(padj < 0.01,log2FoldChange >= 1,baseMean >= 20) %>%
    arrange(desc(log2FoldChange)) %>%
    slice_head(n = 10) 
  
  #This second pipeline both selects the downregulated, binds both
  #and extracts the rownames
  
  rownames_top20 <- as.data.frame(results) %>%
    rename_with(function(x) gsub(".{8}\\.","",x)) %>%
    filter(padj < 0.01,log2FoldChange <= -1,baseMean >= 20) %>%
    arrange(log2FoldChange) %>%
    slice_head(n = 10) %>%
    bind_rows(.,sig_results_up) %>%
    rownames_to_column(var = "rowname") %>%
    pull("rowname")
  
  dds %>% 
    rlog() %>%
    assay() %>%
    as.data.frame() %>%
    filter(row.names(.) %in% rownames_top20) %>%
    pheatmap(.,scale = "row",show_rownames = T,
         clustering_distance_rows = "correlation",
         annotation_col = as.data.frame(colData(dds)[,"Response",drop = F]),
         annotation_row = 
           data.frame(Direction = c(rep("Under-expressed",10), rep("Over-expressed",10))) %>%
           mutate(names = rownames_top20) %>%
           column_to_rownames(var = "names"),
         main = paste(gsub("standard error: Response ","",results@elementMetadata@listData$description[3]),": top 20 DEGs"))
    
}

par(mfrow = c(1,3))
for (i in 1:3){
  results2heatmap(results_contrasts[[i]],dds)
}

#lapply(results_contrasts,function(i) results2heatmap(i,dds))


```

**CR vs PD**

All CR patients cluster together but PD and PR ones don't (please note that of the four PR samples that seem to cluster together, each pair comes from a different branch), which fits what we inferred from the Venn diagram: CR features a lot of exclusive genes which don't differentiate PR from PD.

**PR vs PD**

Most PD patients cluster together, with most of the PR patients clustering with some of the CR ones.Two of the CR patients cluster with the PD patients.

**CR vs PR**

All CR patients cluster in the same top-level branch together with just a few PD patients, which fits what we inferred from the Venn diagram: CR features a lot of exclusive genes which don't differentiate PR from PD.

## GSEA

Now we will check for enriched gene sets in each one of the three comparisons. For that purpose we will use two collections of curated gene sets: the KEGG (Kyoto Encyclopedia of Genes and Genomes) database and the extra-cellular matrix related gene sets produced by [@naba-2012]. We will plot the top 5 up-enriched and the top 5 down-enriched gene sets/pathways in each comparison.

```{r}
curated <- msigdbr(species = "Homo sapiens",category = "C2") %>%
  filter(gs_subcat == "CP:KEGG" | gs_subcat == "CP")
  
curated_use <- curated %>%
  distinct(gs_name,gene_symbol) %>%
  as.data.frame()
```

```{r}
 serial_gsea <- function(results,term2gene) {
  vector_gsea <- as.data.frame(results) %>%
  rename_with(function(x) gsub(".{8}\\.","",x)) %>%
  drop_na(stat) %>%
  arrange(desc(stat)) %>%
  rownames_to_column("Symbol") %>%
  pull(stat,name = Symbol)
  
  over_ct <- GSEA(vector_gsea,TERM2GENE = term2gene) %>%
  pairwise_termsim()
  
  return(over_ct)
}

gsea_contrasts <- lapply(results_contrasts,function(i) serial_gsea(i,curated_use))

par(mfrow = c(1,3))
lapply(1:3,function(i) dotplot(gsea_contrasts[[i]], showCategory = 5,split = ".sign",font.size = 10,label_format = 26,
title = names(gsea_contrasts)[i] %>%
  gsub("res","",.)) +
                              facet_grid(.~.sign))

```

**CR vs PD**

Only one gene set was enriched in the CR patients. Of the five down-enriched ones, four are related to the extra-cellular matrix (the NABA ones). Hedgehog signaling (not shown in the graph, also down-enriched), is key for basal cell carcinoma.

**PR vs PD**

At least three of the top 5 down-enriched pathways (the three NABA ones) are clearly related to the ECM. The complement and coagulation cascades are two different pathways comprised of different sets of genes, but are lumped into one gene set in the KEGG collection. In order to ensure that both are indeed enriched we coloured the enriched genes in the KEGG website, which resulted in extensive colouring of both genes in the complement and in the coagulation cascade.

**CR vs PR**

Here we start to see an interesting and somewhat unexpected pattern: gene sets that are down-enriched in PR with respect to PD are enriched for CR with respect to PR and vice-versa. This will be expanded upon below.

Now let's check the degree of overlap between the gene sets enriched in the different comparisons.

```{r}
lapply(gsea_contrasts,function(i) i[,1]) %>%
venn()
title(main = "Overlap between enriched gene sets")
```

Judging from what we saw in the Venn diagrams of differentially expressed genes, the overlap between the gene sets enriched for B (PR vs PD) and C (CR vs PR) is due to some gene sets being enriched in the same direction for CR and PD respectively to PR. Let's check whether that's the case using the same procedure we used previously:

```{r}
lapply(gsea_contrasts,function(i) i %>% 
         filter(NES < 0) %>%
         as.data.frame() %>%
         pull(ID)) %>%
  venn()
title(main = "Down-enriched gene sets")
```

```{r}
lapply(gsea_contrasts,function(i) i %>% 
         filter(NES > 0) %>%
         as.data.frame() %>%
         pull(ID)) %>%
  venn()
title(main = "Up-enriched gene sets")
```

It is confirmed that the overlap between B and C is due to the above cause. Let's see which are the 29 gene sets enriched in the same direction for CR and PD respectively to PR. The NES shown corresponds to the PR/PD comparison (and therefore is of the opposite sign for the CR/PR comparison).

```{r,results='markup'}
gsea_contrasts[[2]][gsea_contrasts[[2]][,1] %in% gsea_contrasts[[3]][,1],c(1,5)] %>%
  arrange(desc(abs(NES))) %>% 
  dplyr::select(NES) %>%
  datatable()
```

Of these 29 gene sets, 15 are enriched in CR and PD with respect to PR and 14 are enriched in PR with respect to CR and PD. Of these 15, at least 7 have to do with the extra-cellular matrix.

# Discussion

A rigid ECM has been associated with poor response to immunotherapy, probably because of matrix stiffness constituting a physical barrier to T cell infiltration [@he-2021]. Our results show ECM-related gene sets to be enriched both in CR and PR with respect to CR patients. Interestingly, ECM-related genes are enriched in PR with respect to PD, suggesting a PR \< CR \< PD ordering for ECM density, contrasting with the CR \< PR \< PD ordering that would be expected were the difference between PR and CR to be a result of a higher degree of presence in CR of the factors present in PR. Thus, it seems that while a high ECM density is associated with no response, the relationship between ECM and response is not monotonic and too low ECM density can hamper complete responsiveness.

Another gene set whose enrichment is ordered differently to the expected (given the aforementioned assumption) order CR \< PR \< PD is the PPAR signaling pathway, which presents a PR \< CR = PD ordering. The PPAR-gamma signaling pathway is involved in adipogenesis, adipocytes having been linked [@wu-2020] to increased expression of PD-L1 (thus potentially decreasing the effectiveness of pembrolizumab). That ordering may be due to low PPAR signalling levels enabling a limited response while difficulty complete response. The fact that obesity has been linked [@wang-2018] to better response to checkpoint blockade in mice and humans shows that the relationship between adipose tissue and checkpoint blockade response is not straightforward.

Other pathways follow more expected orderings: the hedgehog signalling pathway is enriched both in CR and in PR with respect to PD, but is not significantly enriched in CR with respect to PR (i.e., an CR = PR \< PD ordering). It is possible that HH signalling reduces the effectiveness of pembrolizumab by increasing the expression of PD-L1, which it is known to do [@chakrabarti-2018].

We have to take with caution any conclusion taken from this dataset: the absence of a control group receiving an alternative treatment means we can't fully distinguish between general markers of positive cancer prognosis and specific markers and factors that influence response to pembrolizumab checkpoint blockade.

# Bibliography
